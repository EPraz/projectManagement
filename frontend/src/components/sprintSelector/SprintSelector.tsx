import { memo } from "react";
import { Autocomplete, ListItemIcon } from "@mui/material";
import { LayoutContextProps } from "../../types";
import { useOutletContext } from "react-router-dom";
import DeleteIcon from "@mui/icons-material/Delete";
import AddIcon from "@mui/icons-material/Add";
import {
  DeleteButton,
  SelectorContainer,
  SprintIcon,
  SprintMenuItem,
  SprintNameContainer,
  StyledTextField,
} from "./SprintSelector.styles";

const SprintSelector = () => {
  const {
    loadingDeleteSprint,
    handleChangeSprint,
    listOfSprints,
    setSelectedSprint,
    setOpenCreateSprintDialog,
    setOpenDeleteSprintDialog,
    sprint,
  } = useOutletContext<LayoutContextProps>();

  return (
    <SelectorContainer>
      <SprintIcon />
      <Autocomplete
        value={sprint ? sprint : undefined}
        onChange={(_, newValue) => {
          if (newValue && newValue.id !== "new") {
            handleChangeSprint(newValue.id);
          }
        }}
        // onChange={(_, newValue) => handleChangeSprint(newValue.id)}
        options={[...(listOfSprints || []), { id: "new", name: "New Sprint" }]}
        getOptionLabel={(option) => option?.name}
        isOptionEqualToValue={(option, value) => option.id === value.id}
        disableClearable
        disabled={!listOfSprints?.length}
        renderInput={(params) => (
          <StyledTextField
            {...params}
            placeholder={
              listOfSprints?.length ? "Select Sprint" : "No Sprints Available"
            }
            size="small"
          />
        )}
        renderOption={(props, option) => {
          if (option.id === "new") {
            return (
              <SprintMenuItem
                {...props}
                key={option.id}
                className="new-sprint"
                onClick={(e) => {
                  e.stopPropagation(); // Evita seleccionar la opciÃ³n de "New Sprint"
                  setOpenCreateSprintDialog(true);
                }}
              >
                <ListItemIcon>
                  <AddIcon sx={{ fontSize: 18 }} />
                </ListItemIcon>
                New Sprint
              </SprintMenuItem>
            );
          }

          return (
            <SprintMenuItem {...props} key={`${option.id}_${option.name}`}>
              <SprintNameContainer>{option.name}</SprintNameContainer>
              {"projectId" in option && (
                <DeleteButton
                  size="small"
                  data-testid={"delete-button"}
                  onClick={(e) => {
                    e.stopPropagation(); // Evita cerrar el Autocomplete al eliminar un sprint
                    setSelectedSprint(option);
                    setOpenDeleteSprintDialog(true);
                  }}
                  disabled={loadingDeleteSprint}
                >
                  <DeleteIcon fontSize="small" color="error" />
                </DeleteButton>
              )}
            </SprintMenuItem>
          );
        }}
        sx={{
          width: 200,
          "& .MuiOutlinedInput-root": {
            height: 32,
            fontSize: "0.875rem",
            backgroundColor: "transparent",
            "& fieldset": {
              border: "none",
            },
            "&:hover fieldset": {
              border: "none",
            },
            "&.Mui-focused fieldset": {
              border: "none",
            },
            "& .MuiOutlinedInput-input": {
              padding: 1,
            },
          },
          "& .MuiAutocomplete-endAdornment": {
            display: "none",
          },
        }}
        slotProps={{
          paper: {
            elevation: 4,
            sx: {
              mt: 0.5,
              "& .MuiAutocomplete-listbox": {
                p: 0.5,
                "& .MuiAutocomplete-option": {
                  p: 0,
                },
              },
            },
          },
        }}
      />
    </SelectorContainer>
  );
};

export default memo(SprintSelector);
