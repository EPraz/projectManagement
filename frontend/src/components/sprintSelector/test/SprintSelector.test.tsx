import { render, screen, fireEvent } from "@testing-library/react";
import SprintSelector from "../SprintSelector";
import { MemoryRouter } from "react-router-dom";

// Declaramos mocks para las funciones del contexto que se usan en SprintSelector
const handleChangeSprintMock = jest.fn();
const setSelectedSprintMock = jest.fn();
const setOpenCreateSprintDialogMock = jest.fn();
const setOpenDeleteSprintDialogMock = jest.fn();

// Creamos un mock para el contexto que se espera recibir de useOutletContext
const listOfSprintsMock = [
  { id: "1", name: "Sprint 1", projectId: "123" },
  { id: "2", name: "Sprint 2", projectId: "123" },
];
const sprintMock = { id: "1", name: "Sprint 1", projectId: "123" };

// Mocker useOutletContext para devolver el contexto simulado
jest.mock("react-router-dom", () => {
  const actual = jest.requireActual("react-router-dom");
  return {
    ...actual,
    useOutletContext: () => ({
      loadingDeleteSprint: false,
      handleChangeSprint: handleChangeSprintMock,
      listOfSprints: listOfSprintsMock,
      setSelectedSprint: setSelectedSprintMock,
      setOpenCreateSprintDialog: setOpenCreateSprintDialogMock,
      setOpenDeleteSprintDialog: setOpenDeleteSprintDialogMock,
      sprint: sprintMock,
    }),
  };
});

describe("SprintSelector Component", () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  const renderComponent = () =>
    render(
      <MemoryRouter>
        <SprintSelector />
      </MemoryRouter>
    );

  it("match snapshot", () => {
    const { asFragment } = renderComponent();
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders Autocomplete input with correct placeholder", () => {
    renderComponent();
    expect(screen.getByPlaceholderText("Select Sprint")).toBeInTheDocument();
  });

  it("renders 'New Sprint' option", async () => {
    renderComponent();
    // Simula clic en el input del Autocomplete para abrir la lista de opciones
    const input = screen.getByPlaceholderText("Select Sprint");
    fireEvent.mouseDown(input);
    // Espera a que se renderice la opción "New Sprint"
    const newSprintOption = await screen.findByText("New Sprint");
    expect(newSprintOption).toBeInTheDocument();
  });

  it("calls setOpenCreateSprintDialog when 'New Sprint' option is clicked", async () => {
    renderComponent();
    const input = screen.getByPlaceholderText("Select Sprint");
    fireEvent.mouseDown(input);
    // Encuentra la opción "New Sprint"
    const newSprintOption = await screen.findByText("New Sprint");
    // Simula clic en la opción "New Sprint"
    fireEvent.click(newSprintOption);
    expect(setOpenCreateSprintDialogMock).toHaveBeenCalledWith(true);
    // No se debe llamar handleChangeSprint
    expect(handleChangeSprintMock).not.toHaveBeenCalled();
  });

  it("calls handleChangeSprint when selecting an existing sprint", async () => {
    renderComponent();
    const input = screen.getByPlaceholderText("Select Sprint");
    fireEvent.mouseDown(input);
    // Supongamos que buscamos "Sprint 2"
    const sprintOption = await screen.findByText("Sprint 2");
    fireEvent.click(sprintOption);
    // Se espera que se llame handleChangeSprint con el id "2"
    expect(handleChangeSprintMock).toHaveBeenCalledWith("2");
  });

  it("calls setSelectedSprint and setOpenDeleteSprintDialog when delete button for Sprint 1 is clicked", async () => {
    renderComponent();
    const input = screen.getByPlaceholderText("Select Sprint");
    fireEvent.mouseDown(input);
    // Espera a que se muestren las opciones del Autocomplete
    const sprint1Option = await screen.findByText("Sprint 1");
    expect(sprint1Option).toBeInTheDocument();

    // Obtener todos los botones de eliminación
    const deleteButtons = screen.getAllByTestId("delete-button");
    // Filtrar el botón correspondiente a "Sprint 1"
    const sprint1DeleteButton = deleteButtons.find((button) =>
      button.closest("li")?.textContent?.includes("Sprint 1")
    );
    expect(sprint1DeleteButton).toBeDefined();
    // Simular clic en el botón de eliminación
    fireEvent.click(sprint1DeleteButton!);

    expect(setSelectedSprintMock).toHaveBeenCalledWith(
      expect.objectContaining({ id: "1" })
    );
    expect(setOpenDeleteSprintDialogMock).toHaveBeenCalledWith(true);
  });
});
