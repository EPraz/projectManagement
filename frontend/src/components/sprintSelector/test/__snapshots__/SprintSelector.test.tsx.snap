// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SprintSelector Component match snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-1axwh2v"
  >
    <svg
      aria-hidden="true"
      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1gcnoww-MuiSvgIcon-root"
      data-testid="GroupWorkOutlinedIcon"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2m0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8"
      />
      <circle
        cx="8"
        cy="14"
        r="2"
      />
      <circle
        cx="12"
        cy="8"
        r="2"
      />
      <circle
        cx="16"
        cy="14"
        r="2"
      />
    </svg>
    <div
      class="MuiAutocomplete-root MuiAutocomplete-hasPopupIcon css-d8lkat-MuiAutocomplete-root"
    >
      <div
        class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-1h75mgz-MuiFormControl-root-MuiTextField-root"
      >
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-sizeSmall MuiInputBase-adornedEnd MuiAutocomplete-inputRoot css-1n04w30-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <input
            aria-autocomplete="list"
            aria-expanded="false"
            aria-invalid="false"
            autocapitalize="none"
            autocomplete="off"
            class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall MuiInputBase-inputAdornedEnd MuiAutocomplete-input MuiAutocomplete-inputFocused css-13meb6w-MuiInputBase-input-MuiOutlinedInput-input"
            id=":r0:"
            placeholder="Select Sprint"
            role="combobox"
            spellcheck="false"
            type="text"
            value="Sprint 1"
          />
          <div
            class="MuiAutocomplete-endAdornment css-1uhhrmm-MuiAutocomplete-endAdornment"
          >
            <button
              aria-label="Open"
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium MuiAutocomplete-popupIndicator css-2y136s-MuiButtonBase-root-MuiIconButton-root-MuiAutocomplete-popupIndicator"
              id=":r2:"
              tabindex="-1"
              title="Open"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
                data-testid="ArrowDropDownIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M7 10l5 5 5-5z"
                />
              </svg>
            </button>
          </div>
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1ll44ll-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-w4cd9x"
            >
              <span
                aria-hidden="true"
                class="notranslate"
              >
                â€‹
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
